### ISSUE TITLE
Using AntiPatterns to avoid MLOps Mistakes

### ISSUE THREAD

**Issue by pocca2048** [pocca2048/ML-paper-reading#16]
**Conference** : 
**Link** : https://arxiv.org/pdf/2107.00079.pdf
**Authors' Affiliation** : The Bank of New York Mellon
**TL;DR** : we catalog and document numerous antipatterns in financial ML operations (MLOps)

**Summary** :

![image](https://user-images.githubusercontent.com/10275397/125015241-10738380-e0aa-11eb-9598-3a4eca954b3b.png)

# 3 Antipatterns
supervised learning or forecasting context에서 얘기함.

## 3.1 Data Leakage AntiPattern
* Peek-a-Boo AntiPattern. 
* Temporal Leakage AntiPattern
* Oversampling Leakage AntiPattern
* Metrics-from-Beyond AntiPattern

## 3.2 ‘Act Now, Reflect Never’ AntiPattern
배포해놓고 한번도 안보는 경우. 계속 모니터링해야 concept drift를 막을수 있음.

## 3.3 Tuning-under-the-Carpet AntiPattern
hyperparameter의 중요성.  
it is imperative that the part of a learning pipeline concerned with hyper-parameter optimization be explicitly and painstakingly documented so as to be reproducible and easily adaptable

## 3.4 ‘PEST’ AntiPattern
Perceived Empirical SuperioriTy (PEST)
성능이 좋다고 구라치는 논문들에 속지말자. 많은 논문들은 실험 부족, hyperparameter 튜닝 부족으로 베이스라인보다 좋다고 사기치는것.

## 3.5 Bad Credit Assignment AntiPattern
위에꺼랑 마찬가지. 대부분 performance gain은 제시한 방법 덕분이 아닌데, 그거라고 주장하는 경우가 많음.

## 3.6 Grade-your-own-Exam AntiPattern
 HARKing (Hypothesizing After Results are Known) = test set 가지고서 계속 eval하는 행위.

-->  implicit data leakage

## 3.7 Set & Forget AntiPattern
concept drift

## 3.8 ‘Communicate with Ambivalence’ AntiPattern
uncertainty. calibration도 중요.

## 3.9 ‘Data Crisis as a Service’ AntiPattern
data processing 스텝을 확실하게 기록하고 변하는 걸 감지해야 함.

# 5 RECOMMENDATIONS
1. Use AntiPatterns presented here to document a model management process to avoid costly but routine mistakes in model development, deployment, and approval.
2. Use assertions to track data quality across the enterprise. This is crucial since ML models can be so dependent on faulty or noisy data and suitable checks and balances can ensure a safe operating environment for ML algorithms.
3. Document data lineage along with transformations to support creation of ‘audit trails’ so models can be situated back in time and in specific data slices for re-training or re-tuning.
4. Use ensembles to maintain a palette of models including remedial and compensatory pipelines in the event of errors. Track model histories through the lifecycle of an application 
5. Ensure human-in-the-loop operational capability at multiple levels. Use our model presented for rethinking ML deployment from Section 4 as a basis to support interventions and communication opportunities.
